SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Storage/Storage.cs

```diff
@@ -85,11 +86,38 @@ public Task PutLog(string channelName, Log log)
             return AddTaskToQueue(() =>
             {
                 var logJsonString = LogSerializer.Serialize(log);
-                _storageAdapter.Insert(TableName,
-                    new[] { ColumnChannelName, ColumnLogName },
-                    new List<object[]> {
-                        new object[] {channelName, logJsonString}
-                    });
+                var maxSize = _storageAdapter.GetMaxStorageSize();
+                var logSize = Encoding.UTF8.GetBytes(logJsonString).Length;
+                if (maxSize == -1)
+                {
+                    throw new StorageException("Failed to store a log to the database.");
+                }
+                if (maxSize <= logSize)
+                {
+                    throw new StorageException($"Log is too large ({logSize} bytes) to store in database. " +
+                            $"Current maximum database size is {maxSize} bytes.");
+                }
+                var tryInsert = true;
+                while (tryInsert)
+                {
```

Comment:
I have some concerns about speed, when we catch an exception, remove one line and try again in a cycle. Could we get the current log size? Or maybe remove several lines at once, based on comparing new log size and old log size. 
Just thoughts.

Reply:
This part was reworked